subprojects/base-services/src/main/java/org/gradle/internal/reflect/JavaReflectionUtil.java

```diff
@@ -55,4 +63,99 @@
             throw UncheckedException.throwAsUncheckedException(e);
         }
     }
+
+    /**
+     * Checks if a type has a type variable which may require resolving.
+     */
+    public static boolean hasTypeVariable(Type type) {
+        // do some checks up-front, so we avoid creating the queue in most cases
+        // Cases we want to handle:
+        // - List<String>
+        // - Class<?>
+        // - List<Class<?>>
+        // - Integer[]
+        // - ? extends BaseType
+        // - Class<?>[]
+        if (doesNotHaveTypeVariable(type)) {
+            return false;
+        }
+        if (type instanceof TypeVariable) {
+            return true;
+        }
+        if (type instanceof ParameterizedType) {
+            ParameterizedType parameterizedType = (ParameterizedType) type;
+            boolean noTypeVariables = true;
+            for (Type actualTypeArgument : parameterizedType.getActualTypeArguments()) {
+                if (actualTypeArgument instanceof TypeVariable) {
+                    return true;
+                }
+                noTypeVariables &= doesNotHaveTypeVariable(actualTypeArgument);
+            }
+            if (noTypeVariables) {
+                return false;
+            }
+        }
+        if (type instanceof GenericArrayType) {
+            GenericArrayType genericArrayType = (GenericArrayType) type;
+            if (genericArrayType.getGenericComponentType() instanceof TypeVariable) {
+                return true;
+            }
+        }
+
+        // Type is more complicated, need to check everything.
+        Queue<Type> typesToInspect = new ArrayDeque<Type>();
+        typesToInspect.add(type);
+        while (!typesToInspect.isEmpty()) {
+            Type typeToInspect = typesToInspect.remove();
+            if (typeToInspect instanceof Class) {
+                continue;
+            }
+            if (typeToInspect instanceof TypeVariable) {
+                return true;
+            }
+            if (typeToInspect instanceof ParameterizedType) {
+                ParameterizedType parameterizedType = (ParameterizedType) typeToInspect;
+                CollectionUtils.addAll(typesToInspect, parameterizedType.getActualTypeArguments());
+            } else if (typeToInspect instanceof GenericArrayType) {
+                GenericArrayType arrayType = (GenericArrayType) typeToInspect;
+                typesToInspect.add(arrayType.getGenericComponentType());
+            } else if (typeToInspect instanceof WildcardType) {
+                WildcardType wildcardType = (WildcardType) typeToInspect;
+                CollectionUtils.addAll(typesToInspect, wildcardType.getLowerBounds());
+                CollectionUtils.addAll(typesToInspect, wildcardType.getUpperBounds());
+            } else {
+                // We don't know what the type is - let Guava take care of it.
+                return true;
```

Comment:
So the method is only detecting if the type _might_ have a type variable?

Reply:
No, it detects if the type has a type variable. To my knowledge, the above cases are everything. Resolving the type variable is done by TypeToken.

I could also have thrown an exception at this point, I just though I wanted to err on the side of delegating to Guava.

Reply:
Got it. It's still weird that we just return `true` in this case, and the comment doesn't describe what's happening too well for me. I don't have a suggestion about how to make it better, though.